rerun_cpp/src/rerun/component_batch.hpp

```diff
@@ -1,94 +1,336 @@
+        static ComponentBatch<TComponent> take_ownership(TComponent&& data) {
+            // TODO(andreas): there should be a special path here to avoid allocating a vector.
+            return take_ownership(std::vector<TComponent>{std::move(data)});
+        }
 
-      public:
-        /// Construct from any parameter that can be converted to a strongly typed component list.
-        template <typename ComponentBatchLikeType>
-        AnonymousComponentBatch(const ComponentBatchLikeType& component_list_like)
-            : AnonymousComponentBatch(ComponentBatch(component_list_like)) {}
-
-        /// Construct from a strongly typed component list.
-        template <typename ComponentType>
-        AnonymousComponentBatch(const ComponentBatch<ComponentType>& component_list)
-            : data(component_list.data),
-              num_instances(component_list.num_instances),
-              to_data_cell_func([](const void* _data, size_t _num_instances) {
-                  return ComponentType::to_data_cell(
-                      reinterpret_cast<const ComponentType*>(_data),
-                      _num_instances
-                  );
-              }) {}
-
-        /// Creates a Rerun DataCell from this list of components.
-        Result<rerun::DataCell> to_data_cell() const {
-            return to_data_cell_func(data, num_instances);
+        /// Move constructor.
+        ComponentBatch(ComponentBatch<TComponent>&& other) : ComponentBatch() {
+            *this = std::move(other);
+        }
+
+        /// Move assignment
+        void operator=(ComponentBatch<TComponent>&& other) {
+            // Perform a swap with `other`.
+            // (writing out this-> here to make it less confusing!)
+            switch (this->ownership) {
+                case BatchOwnership::Borrowed: {
+                    switch (other.ownership) {
+                        case BatchOwnership::Borrowed:
+                            std::swap(this->storage.borrowed, other.storage.borrowed);
+                            break;
+
+                        case BatchOwnership::VectorOwned: {
+                            auto this_borrowed_data_old = this->storage.borrowed.data;
+                            this->storage.vector_owned = std::move(other.storage.vector_owned);
+                            other.storage.borrowed.data = this_borrowed_data_old;
+                            break;
+                        }
+                    }
+                    break;
+                }
+
+                case BatchOwnership::VectorOwned: {
+                    switch (other.ownership) {
+                        case BatchOwnership::Borrowed: {
+                            auto other_borrowed_data_old = other.storage.borrowed.data;
+                            other.storage.vector_owned = std::move(this->storage.vector_owned);
+                            this->storage.borrowed.data = other_borrowed_data_old;
+                            break;
```

Comment:
⚠️ You have forgotten about `num_instances` !!!

Reply:
woah. good catch. Need to add a test that catches it